### PR TITLE
[New post] How to build a VTubing app with Amazon Interactive Video Service and VRoid

### PR DESCRIPTION
**Pull request by tonyv**
…eo Service and VRoid

# Content Pull Request Checklist

Please check to make sure your content submission meets the following requirements by adding an `x` between the brackets:

- [X] The title of this pull request reflects the title of my content.
- [X] I have reviewed the [Content Review Checklist](https://github.com/build-on-aws/content/blob/main/CONTENT_REVIEW_CHECKLIST.md).
- [X] This pull request contains one piece of content (post or page).

**Important: if you are submitting a PR to update existing content, please ensure you [sync'ed your fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork) first before you created your branch.**

Is this pull request linked to a Content Proposal or Idea Suggestion Issue? If so, please link to it below using the number or full link.

```text
Resolves #5

or

Content for https://github.com/build-on-aws/content/issues/5
```

**By submitting this pull request, I confirm I own this content, have permission to use each image included in this content, and that it will be released under the [CC BY-SA 4.0 SA License](https://github.com/build-on-aws/content/blob/main/LICENSE). Notwithstanding the above, if you provide source code text within your content piece, you are providing such source code under the [MIT-0](https://opensource.org/license/mit-0/) license.**

### CODE REVIEW

File: tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+| ✅ AWS Level        | Intermediate - 200                                                                                                             |
+| ⏱ Time to complete  | 15 minutes                                                                                                                     |
+| 💰 Cost to complete | Free when using the AWS Free Tier or USD 1.01                                                                                  |
+| 🧩 Prerequisites    | - [AWS Account](https://aws.amazon.com/resources/create-account/)<br>- [Pixiv Account](https://accounts.pixiv.net/signup) <br> |
+| 💻 Code Sample      | Code sample used in tutorial on [GitHub](#)                                                                                    |
+| 📢 Feedback         | <a href="https://pulse.buildon.aws/survey/DEM0H5VW" target="_blank">Any feedback, issues, or just a</a> 👍 / 👎 ?              |
+| ⏰ Last Updated     | YYYY-MM-DD                                                                                                                     |
```
Comment (darko-mesaros) [inline tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:43]: Can you just update the date here.

File: tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+- How to live stream your virtual character to Amazon IVS
+
+| Attributes          |                                                                                                                                |
+| ------------------- | ------------------------------------------------------------------------------------------------------------------------------ |
+| ✅ AWS Level        | Intermediate - 200                                                                                                             |
+| ⏱ Time to complete  | 15 minutes                                                                                                                     |
+| 💰 Cost to complete | Free when using the AWS Free Tier or USD 1.01                                                                                  |
+| 🧩 Prerequisites    | - [AWS Account](https://aws.amazon.com/resources/create-account/)<br>- [Pixiv Account](https://accounts.pixiv.net/signup) <br> |
+| 💻 Code Sample      | Code sample used in tutorial on [GitHub](#)                                                                                    |
```
Comment (darko-mesaros) [inline tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:41]: Add the github link to the code repo here.

File: tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+
+## Part 1 - Integrating with the VRoid Hub API and loading an avatar
+
+First, [sign up for a Pixiv account](https://accounts.pixiv.net/signup) and follow the [VRoid Hub quick start guide](https://developer.vroid.com/en/api/quick-start.html) to create an application. Once you create an application, you can reference the [code sample](https://github.com/pixiv/VRoidHub-API-Example) provided by Pixiv that demonstrates how to download a character on the VRoid Hub as a VRM file. VRM is a file format for handling 3D character models. For this tutorial, we have created our own 3D character using [VRoid Studio](https://vroid.com/en/studio). VRoid Studio is a 3D character creator tool that lets you export VRM files locally or upload them to VRoid Hub to share with the public. After creating our 3D character, we have saved it in VRM format and made it available [here](https://d1l5n2avb89axj.cloudfront.net/avatar-first.vrm) for use in this tutorial and [on VRoid Hub here](https://hub.vroid.com/en/characters/1692004541821223967/models/6410417122824343972).
```
Comment (darko-mesaros) [inline tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:61]: I'd love to see some additional instructions here, as the VRoid quick start guide was a bit confusing (few items in the new application wizard I just made up). Just to make it a bit smoother to the reader, you should anticipate this confusion for people who have never used VRoid

Comment (darko-mesaros) [reply tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:61]: The code sample is leading us to a Repo that is mostly in Japanese 😓 I was not sure if I should read it and follow instructions. If this is not necessary to proceed with the tutorial I suggest you move it to a little Indented segment:
> Like this
And then continue with the steps needed - and that is to download the VRM file itself+

File: tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+
+```html
+<script src="https://cdn.jsdelivr.net/npm/@mediapipe/holistic@0.5.1635989137/holistic.js" crossorigin="anonymous"></script>
+<script src="https://cdn.jsdelivr.net/npm/kalidokit@1.1/dist/kalidokit.umd.js"></script>
+```
+
+Now let’s add code for our character rigging logic and create the animateVRM function. The animateVRM function will be used to get landmark data from MediaPipe Holistic passed in via the results argument. In the context of rigging, landmark data refers to specific points of a body part. Using the landmark data from MediaPipe Holistic, we can pass it to Kalidokit to animate specific body parts. You can view the complete function for doing the animation [here]().
```
Comment (darko-mesaros) [inline tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:120]: Missing link to [here]

File: tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+```html
+<script src="https://unpkg.com/three@0.133.0/build/three.js"></script>
+<script src="https://unpkg.com/three@0.133.0/examples/js/loaders/GLTFLoader.js"></script>
+<script src="https://unpkg.com/@pixiv/three-vrm@0.6.7/lib/three-vrm.js"></script>
+<script src="https://unpkg.com/three@0.133.0/examples/js/controls/OrbitControls.js"></script>
+```
+
+After importing these libraries, let’s use the Three.js library to load a [scene object](https://threejs.org/docs/#api/en/scenes/Scene). A scene is used in Three.js to load in our 3D character or avatar. You will then use the [WebGLRenderer](https://threejs.org/docs/#api/en/renderers/WebGLRenderer) provided by Three.js to to actually display the scene. Using the renderer object created from WebGLRenderer, we dynamically add a canvas element to our HTML. This canvas element will be used to render the scene and our avatar.
+
+```javascript
```
Comment (darko-mesaros) [inline tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:78]: Where is this code supposed to go? The `index.html` file?

File: tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+
+Virtual YouTubing, or V-Tubing for short, is the practice among live streamers of using virtual avatars to either present their brand, their personality, or their identity through means outside of a face cam.
+
+With that context set, I will walk you through a step by step process of building a web app to load a 3D character from VRoid Hub, animate it based on your body movements, and live stream to Amazon Interactive Video Service (IVS). [VRoid Hub](https://hub.vroid.com/en) is an online platform created by Pixiv used to host and share 3D character models. You will be using the VRoid Hub API to import a digital character from VRoid Hub and use it as your avatar. The Three VRM SDK from Pixiv will be used for rendering the digital character. To have your avatar animate and mimic your body movements, you will be using the MediaPipe Holistic library and [Kalidokit](https://github.com/yeemachine/kalidokit).
+
+> The code for this tutorial is available on [Github](#). You can also try a [live demo](https://dev.d218eir1ybnzul.amplifyapp.com/) of this web app.
```
Comment (darko-mesaros) [inline tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:27]: Missing the github link here.

File: tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+
+  (error) => console.error(error)
+);
+```
+
+## Part 3 - Animating a virtual character with your own body movements
+
+To start animating your virtual character, install the [Kalidokit](https://github.com/yeemachine/kalidokit) and [MediaPipe Holistic](https://github.com/google/mediapipe/blob/master/docs/solutions/holistic.md) libraries. MediaPipe Holistic is a computer vision pipeline used to track a user’s body movements, facial expressions, and hand gestures. This is useful for animating your digital avatar to mimic your own movements. Kalidokit includes the use of blendshapes for facial animation and kinematics solvers for body movements to create more realistic digital avatars. Blendshapes are a technique used in character animation to create a wide range of facial animations. Kinematics solvers are algorithms used to calculate the position and orientation of an avatar’s limbs. When making our avatar animate, aka character rigging, a kinematics solver helps determine how a character’s joints and bones should move to achieve a desired pose or animation. In short, Mediapipe Holistic tracks your physical movements while Kalidokit takes those as inputs to animate your avatar.
```
Comment (darko-mesaros) [inline tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:113]: You are using the term "Install" here but I am not sure if there is any installation to be made. If we are just importing them to the HTML there should be nothing to install. I'd suggest you change the word from "install" to "import" or something similar :)

**Comment by darko-mesaros** [comment]
OVERALL FEEDBACK: I could not finish this tutorial, I believe critical parts of it are missing. Namely the JavaScript code that was presented, I am not really sure where to write that as adding it too the `index.html` file makes no sense at least in the way it's presented. 

The goal of this tutorial is to have any one be able to pick it up and wind up with the same results as you had! So, I'd love if you would be a bit more granular with the steps needed to get there. Thank you!

**Comment by tonyv** [comment]
@darko-mesaros Very much appreciate the thorough review and feedback. I've addressed all feedback and just waiting for links to the Github repo referenced in the blog before pushing another commit.

**Comment by tonyv** [comment]
Hi @darko-mesaros - Quick update that I'm still waiting on the Github repo links. Would you be able to review the other changes I have made in the meantime?

**Review by cherbk — APPROVED** [review]

File: tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md [hunk 8]
```diff
@@ -0,0 +1,540 @@
...
+
+      <input type="text" id="participant-token" placeholder="Enter a participant token"  />
+      <button class="btn" id="submit-btn">Submit</button>
+    </section>
+    <div class="overlay hidden"></div>
+  </body>
+</html>
+```
+
+## Part 3 - Rendering a virtual character with the three VRM SDK
```
Comment (DalM2023) [inline tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:104]:
```suggestion
## Part 3 - Rendering a Virtual Character With the Three VRM SDK
```

File: tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md [hunk 9]
```diff
@@ -0,0 +1,540 @@
...
...ster/docs/solutions/holistic.md) library, and camera utility module from MediaPipe to the `<head>` element in `index.html`. MediaPipe Holistic is a computer vision pipeline used to track a user’s body movements, facial expressions, and hand gestures. This is useful for animating your digital avatar to mimic your own movements. Kalidokit includes the use of blendshapes for facial animation and kinematics solvers for body movements to create more realistic digital avatars. Blendshapes are a technique used in character animation to create a wide range of facial animations. Kinematics solvers are algorithms used to calculate the position and orientation of an avatar’s limbs. When making our avatar animate, aka character rigging, a kinematics solver helps determine how a character’s joints and bones should move to achieve a desired pose or animation. In short, MediaPipe Holistic tracks your physical movements while Kalidokit takes those as inputs to animate your avatar. The camera utility module from MediaPipe will simplify the process of providing our front facing camera input to MediaPipe Holistic. MediaPipe Holistic needs this camera input to do hand, face and body movement tracking.
```
Comment (DalM2023) [inline tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:185]:
```suggestion
To start animating your virtual character, add `<script>` elements for the [Kalidokit](https://github.com/yeemachine/kalidokit) library, [MediaPipe Holistic](https://github.com/google/mediapipe/blob/master/docs/solutions/holistic.md) library, and camera utility module from MediaPipe to the `<head>` element in `index.html`. MediaPipe Holistic is a computer vision pipeline used to track a user’s body movements, facial expressions, and hand gestures. This is useful for animating your digital avatar to mimic your own movements. Kalidokit includes the use of blendshapes for facial animation and kinematics solvers for body movements to create more realistic digital avatars. Blendshapes are a technique used in character animation to create a wide range of facial animations. Kinematics solvers are algorithms used to calculate the position and orientation of an avatar’s limbs. When making our avatar animate, aka character rigging, a kinematics solver helps determine how a character’s joints and bones should move to achieve a desired pose or animation. In short, MediaPipe Holistic tracks your physical movements while Kalidokit takes those as inputs to animate your avatar. The camera utility module from MediaPipe will simplify the process of providing our front-facing camera input to MediaPipe Holistic. MediaPipe Holistic needs this camera input to do hand, face and body movement tracking.
```

File: tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md [hunk 10]
```diff
@@ -0,0 +1,540 @@
...
+  onFrame: async () => {
+    await holistic.send({ image: videoElement });
+  },
+  width: 640,
+  height: 480,
+});
+camera.start();
+```
+
+## Part 5 - Live stream your virtual character to Amazon IVS
```
Comment (DalM2023) [inline tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:407]:
```suggestion
## Part 5 - Live Stream Your Virtual Character to Amazon IVS
```

File: tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md [hunk 11]
```diff
@@ -0,0 +1,540 @@
...
+
+To publish our video so the audience can see it, let’s capture the MediaStream from our canvas element and assign it to avatarStream in a function called init(). We use the captureStream method of the Canvas API to do so.
+
+```javascript
+const init = async () => {
+  const avatarStream = renderer.domElement.captureStream();
+};
+```
+
+Once we have a MediaStream we want to publish to an audience, we need to join a stage. Joining a stage enables us to live stream the video feed to the audience or other participants in the stage. If we don’t want to live stream anymore, we can leave the stage. Let’s add event listeners that listen for click events when an end user clicks the join or leave stage buttons and implement the appropriate logic.
```
Comment (DalM2023) [inline tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:423]:
```suggestion
Once we have a MediaStream we want to publish to an audience and we need to join a stage. Joining a stage enables us to live stream the video feed to the audience or other participants in the stage. If we don’t want to live stream anymore, we can leave the stage. Let’s add event listeners that listen for click events when an end user clicks the join or leave stage buttons and implement the appropriate logic.
```

File: tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md [hunk 12]
```diff
@@ -0,0 +1,540 @@
...
+    shouldSubscribeToParticipant() {
+      return SubscribeType.AUDIO_VIDEO;
+    },
+  };
+
+  stage = new Stage(token, strategy);
+};
+```
+
+Finally, let’s add some logic to listen for Stage events. These events occur when the state of the stage you’ve joined changes such as when someone joins or leaves it. Using these events, you can dynamically update the HTML code to display a new participant’s video feed when they join or remove it from display when they leave. The setupParticipant and teardownParticipant functions do each of these actions respectively. As a step step we call the join method on the stage object to join the stage.
```
Comment (DalM2023) [inline tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:489]:
```suggestion
Finally, let’s add some logic to listen for Stage events. These events occur when the state of the stage you’ve joined changes such as when someone joins or leaves it. Using these events, you can dynamically update the HTML code to display a new participant’s video feed when they join or remove it from display when they leave. The setupParticipant and teardownParticipant functions do each of these actions respectively. As a step we call the join method on the stage object to join the stage.
```

File: tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md [hunk 13]
```diff
@@ -0,0 +1,540 @@
...
+  console.error(err.message);
+}
+```
+
+At this point, we are now broadcasting our live avatar feed that is mimicking our body movements to a stage. To test if someone else joining the stage can see your avatar, open the [Amazon IVS Real-Time Streaming Web Sample](https://codepen.io/amazon-ivs/project/editor/DYapzL#) in another browser window, create another participant token, provide it in this browser window and click join stage. You should now see the avatar move as you move on camera. Notice the latency can sub-second and can be as low as 300ms. This is how other audience members joining the stage would see your avatar.
+
+## Conclusion
+
+In this tutorial, you created a VTubing app by leveraging Pixiv’s SDKs to display a virtual avatar character and live stream it using Amazon IVS. VTubing opens the doors to an exciting world of virtual content creation. By following the steps outlined in this tutorial, you have gained the knowledge and tools necessary to bring your unique virtual persona to life.
+
```
Comment (DalM2023) [inline tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:535]: Great conclusion! We want to invite readers to stay in the blog repository longer. Please add a sentence or two that is apropos of your topic with 2-3 blog links.  

EX: "To read more/learn more about X, visit (LINK) and (LINK)."

File: tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md [hunk 14]
```diff
@@ -0,0 +1,540 @@
+---
+title: "How to build a VTubing app with Amazon Interactive Video Service and VRoid"
+description: "A step by step guide to set up a 3D virtual avatar that follows your body movements and leverage live streaming from Amazon IVS to broadcast to a large audience."
+tags:
```
Comment (DalM2023) [inline tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:4]: We can only use 5 tags. Please choose the best ones.

File: tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md [hunk 15]
```diff
@@ -0,0 +1,540 @@
+---
+title: "How to build a VTubing app with Amazon Interactive Video Service and VRoid"
+description: "A step by step guide to set up a 3D virtual avatar that follows your body movements and leverage live streaming from Amazon IVS to broadcast to a large audience."
```
Comment (DalM2023) [inline tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:3]: The descriptions need to be 60 characters or fewer. They can't repeat the messaging in the title or use the same words. They should focus on what solution the reader gets from reading the piece. I wrote an example below but feel free to rewrite. 

```suggestion
description: "Make a 3D avatar with live streaming to broadcast to a large audience."
```

File: tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md [hunk 16]
```diff
@@ -0,0 +1,540 @@
+---
+title: "How to build a VTubing app with Amazon Interactive Video Service and VRoid"
```
Comment (DalM2023) [inline tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:2]:
```suggestion
title: "How To Build a VTubing App With Amazon Interactive Video Service and VRoid"
```

File: tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md [hunk 17]
```diff
@@ -0,0 +1,540 @@
+---
```
Comment (DalM2023) [inline tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:1]: Hello, 


What a great opening line and what a fun project! 

Hello,

What a fun project! 

This is a well written blog with good structure and sources.  My edits focus primarily on correcting titles to AP Style and proper name capitalization errors. Please see my notes below. 

Overall, well done!

**Review by cherbk — APPROVED** [review]

**Review by cherbk — APPROVED** [review]

**Comment by cherbk** [comment]
@damonk10 Please do final review but not publish. The author has requested this piece be published in mid  January. Thanks

**Comment by tonyv** [comment]
> @damonk10 Please do final review but not publish. The author has requested this piece be published in mid January. Thanks

@damonk10 @cherbk Kindly following up on this. If possible, could this be published around Jan 15?

**Review by cherbk — APPROVED** [review]

**Review by cherbk — APPROVED** [review]

**Pull request merged**